source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,405 @@ namespace EnvoyDefault {
+  // https://www.rfc-editor.org/rfc/rfc9110.html#section-15:
+  //
+  // The status code of a response is a three-digit integer code that describes the result of the
+  // request and the semantics of the response, including whether the request was successful and
+  // what content is enclosed (if any). All valid status codes are within the range of 100 to 599,
+  // inclusive.
+
+  static uint32_t kMinimumResponseStatusCode = 100;
+  static uint32_t kMaximumResponseStatusCode = 999;
+  const auto& value_string_view = value.getStringView();
+
+  // Convert the status to an integer.
+  std::uint32_t status_value{};
+  auto result = std::from_chars(value_string_view.begin(), value_string_view.end(), status_value);
+  if (result.ec != std::errc() || result.ptr != value_string_view.end()) {
+    return {HeaderValueValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidStatus};
+  }
+
+  if (status_value < kMinimumResponseStatusCode || status_value > kMaximumResponseStatusCode) {
+    return {HeaderValueValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidStatus};
+  }
+
+  return HeaderValueValidationResult::success();
 }
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateResponseHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateGenericHeaderName(const HeaderString& name) {
+  // Verify that the header name is valid. This also honors the underscore in
+  // header configuration setting.
+  //
+  // From RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-5.1:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-name     = token
+  // token          = 1*tchar
+  //
+  // tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
+  //                / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
+  //                / DIGIT / ALPHA
+  //                ; any VCHAR, except delimiters
+  const auto& key_string_view = name.getStringView();
+  bool allow_underscores =
+      config_.headers_with_underscores_action() == HeaderValidatorConfig::ALLOW;
+  // This header name is initially invalid if the name is empty.
+  if (key_string_view.empty()) {
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  bool is_valid = true;
+  char c = '\0';
+
+  for (auto iter = key_string_view.begin(); iter != key_string_view.end() && is_valid; ++iter) {
+    c = *iter;
+    is_valid &= testChar(kGenericHeaderNameCharTable, c) && (c != '_' || allow_underscores);
+  }
+
+  if (!is_valid && c == '_' &&
+      config_.headers_with_underscores_action() == HeaderValidatorConfig::DROP_HEADER) {
+    return {HeaderEntryValidationResult::Action::DropHeader,
+            UhvResponseCodeDetail::get().InvalidUnderscore};
+  } else if (!is_valid) {
+    auto details = c == '_' ? UhvResponseCodeDetail::get().InvalidUnderscore
```

Comment:
I'd definitely be inclined to return invalid characters for all these errors, rather than focus on the underscore disallowed by policy.  Invalid characters is a bigger error.

Reply:
you marked this as resolved but I don't see a push since my last comments?  unresolving

Reply:
It looks like I didn't push after I resolved the discussion.